Make Peer LESProtocol Methods generate own Request ID's

Bhargavasomu opened the pull request:
### What was wrong?
Fixes #27


### How was it fixed?
The argument of `request_id` was made by default `None` in all the `LESProtocol Function Signatures` which use this argument. The main idea here is that the `Peers` don't pass in the `request_id` and the `request_id` would be automatically generated and returned by the `request calls(LESProtocol Methods)` made by them. 

On the other hand the `Server` which accepts these `requests` has to pass the `request_id` which it receives, so that it can create a `reply` with the same `request_id`. Hence the `Server` or `Remote Peer` passes the argument `request_id` to the `LESProtocol Functions` (unlike the peers which generate their own). This functionality of maintaining the `request_id` in the `server` side is needed because it has to send a `reply` with the same `request_id` as the `request` which it receives, so that the `client` or `peer` which receives a `reply`, knows to which `request` it has got a `reply` to.


#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSqtKVVaJ1na9IfqfQK8XX5jdy2ict5YhJJxKftE5xXt9pev3yt)

Bhargavasomu commented:
@pipermerriam I tried writing a basic test as follows
```
@pytest.mark.asyncio
async def test_les_protocol_methods_given_request_id(les_peer_and_remote):
    peer1, peer2 = les_peer_and_remote

    assert isinstance(peer1.sub_proto, LESProtocol)
    assert isinstance(peer2.sub_proto, LESProtocol)

    # Test for get_block_headers
    peer1.sub_proto.send_get_block_bodies([b'123'], request_id=2)
    cmd, msg = await peer2.read_msg()
```
This is throwing some error in the `Decryption of the Header` at `peer2` side. Is there anything I am missing in connecting the 2 peers?

Bhargavasomu commented:
@pipermerriam done. You can see the error in `py37-core`. Thankyou.

pipermerriam commented:
@Bhargavasomu two problems happening here.

First, `LESProtocol` doesn't know how to respond to `GetBlockBodies` so this will fail once the decryption error is resolved.  Converting to using `peer1.sub_proto.send_get_block_headers` will fix this.

Next, the decryption error is happening because the `Peer` classes have an internal message loop which is always reading off incoming messages.  You can fix this by updating your code to use the following API to retrieve the message.

```python
    # Test for get_block_headers
    with peer2.collect_sub_proto_messages() as buffer:
        peer1.sub_proto.send_get_block_headers(b'unicornsrainbowsunicornsrainbows', 1, 0, False)
        await asyncio.sleep(0.1)

    messages = buffer.get_messages()
    assert len(messages) == 1
    cmd, msg = messages[0]

    # assert some things...
```

Lemme know if you hit any other snags.

Bhargavasomu commented:
@pipermerriam I have added the tests for the `block_header`. But I actually wanted to test on all the `LESProtocol Methods`.

> LESProtocol doesn't know how to respond to GetBlockBodies so this will fail

Sorry for being naive here, but why doesn't `LESProtocol` support other commands like `GetBlockBodies`? Is it the case that `LESProtocol` was designed in such a way so as to not support it (If this is the case, then why is this command included as one of the methods in the `LESProtocol`)? Or haven't we added support to all the other commands yet?

pipermerriam commented:
> why doesn't `LESProtocol` support other commands like `GetBlockBodies`?

Take a look at this:

https://github.com/ethereum/trinity/blob/2e8f681818c6befb3ecef0247f9380fba7122156/trinity/protocol/les/servers.py

The `LESProtocol` knows how to *send* a `GetBlockBodies` request but it *does not* know how to respond to one.  In your example `alice` sends the request and when `bob` receives it it doesn't know how to handle it.  The reason for this is that an `LESPeer` won't have access to block bodies because it is a light client.

Currently, we have no constructs for being on the server side of an `LESPeer` connection so we won't be able to easily test things like `GetBlockBodies` requests.  This will require additional work to update our `Peer` classes in a way that allows them to serve the `LES` protocol.

Bhargavasomu commented:
@pipermerriam so to overcome this, if we could somehow make `Alice` a `LESPeer` and `Bob` a `EthPeer` then we should be able to get a response. Is this approach right or am I missing anything?
